File: 백준/Gold/9019. DSLR/DSLR.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

class Main{
    static class RegisterNumber{
        int value;
        StringBuilder path;

        public RegisterNumber(int value, StringBuilder path){
            this.value = value;
            this.path = path;
        }
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine());

        StringBuilder answer = new StringBuilder();
        while(t-- > 0){
            StringTokenizer stk = new StringTokenizer(br.readLine());
            int origin = Integer.parseInt(stk.nextToken());
            int target = Integer.parseInt(stk.nextToken());
            answer.append(bfs(origin, target)).append("\n");
        }

        System.out.print(answer);
    }

    private static String bfs(int origin, int target){
        Queue<RegisterNumber> que = new LinkedList<>();
        boolean[] visit = new boolean[10000];
        final char[] command = new char[]{'D', 'S', 'L', 'R'};

        que.add(new RegisterNumber(origin, new StringBuilder()));

        while(!que.isEmpty()){
            RegisterNumber cur = que.poll();
            
            int D = (cur.value * 2) % 10000;
            int S = cur.value == 0 ? 9999 : cur.value - 1;

            int L = (cur.value % 1000) * 10 + (cur.value / 1000);
            int R = (cur.value % 10) * 1000 + (cur.value / 10);

            int[] numbers = new int[]{D, S, L, R};

            for(int i=0 ; i<4 ; i++){
                int nextValue = numbers[i];

                if(nextValue == target)
                    return cur.path.append(command[i]).toString();

                if(visit[nextValue])continue;

                visit[nextValue] = true;
                que.add(new RegisterNumber(nextValue, new StringBuilder(cur.path).append(command[i])));
            }
        }

        return "";
    }
}

File: 백준/Gold/9019. DSLR/DSLR.py
from collections import deque
from sys import stdin

input = stdin.readline

T = int(input())


def bfs(start, target):
    que = deque()
    que.append((start, ''))
    visit = [False] * 10000
    
    while que:
        cur, path = que.popleft()
        #print(path, cur)
        
        D = (2 * cur) % 10000
        if D == target:
            return path + 'D'
        if not visit[D]:
            visit[D] = True
            que.append((D, path + 'D'))

        S = cur - 1 if cur else 9999
        if S == target:
            return path + 'S'
        if not visit[S]:
            visit[S] = True
            que.append((S, path + 'S'))

        str_cur = str(cur)
        L = int(str_cur[1:] + str_cur[0]) if len(str_cur) == 4 else int(str_cur) * 10
        if L == target:
            return path + 'L'
        if not visit[L]:
            visit[L] = True
            que.append((L, path + 'L'))

        R = int(str_cur[-1] + str_cur[:-1]) if len(str_cur) == 4 else int(str_cur[-1] + '0' * (4 - len(str_cur)) + str_cur[:-1])
        if R == target:
            return path + 'R'
        if not visit[R]:
            visit[R] = True
            que.append((R, path + 'R'))


for _ in range(T):
    s, t = map(int, input().split())
    print(bfs(s, t))


